azure/services/scalesets/scalesets.go

```diff
@@ -369,6 +369,28 @@ func (s *Service) validateSpec(ctx context.Context) error {
 		}
 	}
 
+	// Validate DiagnosticProfile spec
+	if spec.DiagnosticsProfile.Boot != nil {
```

Comment:
This duplicates some of the logic from `ValidateDiagnostics` in `azuremachine_validation.go`. Would it be worth refactoring so that code could be shared somehow?

Reply:
I tried following how validation is done for other features there and I couldn't find cases where this is done as a shared check for both Webhooks and Runtime validation.

Looking at the code, I think it would require a significant amount of logic to make it generic enough to suit both usages also a big part of those lines is around formatting the errors, which is done in differently in Webhooks vs Runtime validation, which is why I think this is done separately for the other features in the first place.

Hence I think it's not really worth it, but I'm happy to commit to it if we think it's essential :)

Reply:
/hold

I don't see AzureMachinePool types being modified anywhere

Reply:
In order to make this field configurable for scale sets (ie. MachinePools), you'll have to also modify MachinePool types, webhook, etc. The AzureMachinePool types are located in the exp/ folder. I see two ways of moving this PR forward:

1) Remove the changes in the scalesets package, make this PR scoped to VMs only and open an issue + follow up PR to make the equivalent changes for VMSS
2) Implement the equivalent changes made for AzureMachine to AzureMachinePool

Reply:
My preference would be for #2, as (most) VM features are common across both, and we want to expose them to both Machine and MachinePool configurations. Doing it in one PR also has the side-benefit of being a one-stop reference for future implementers of such features. This PR is 💯 on the way towards that, thank you @damdo!

Reply:
This is mostly for my own understanding, but is this set of checks required on top of the webhook validation that is happening inside of [`azuremachinepool_webhook.go`](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2528/files#diff-435f85313e217cd2c7e0569bdace226dcb57ba2210dff964a2e511193010f809R196)? It seems like the checks are identical so I'm wondering if there is a specific reason for that.

Reply:
I think it's normally considered good practice to validate both at a webhook level and internally (at runtime).
The first is normally for informing who is applying the config of its invalidity and potentially block the configuration to be persisted in the cluster. The second is for runtime validation before taking action, as webhooks are not always guaranteed to be up and running. So they serve different purposes.
At least this is my understanding of it.

Reply:
> I think it's normally considered good practice to validate both at a webhook level and internally (at runtime)

This is not a convention we follow in this repo. Maintaining duplicate validation logic also has a maintenance cost as both places need to be updated when validation rules change, and there is no test that enforces that they are consistent, especially if we are duplicating code and not sharing common helper functions.

>  as webhooks are not always guaranteed to be up and running

I think it's okay to make the assumption that webhooks should always be up and running. Without webhooks, we don't get defaults either which would result in a lot of invalid configurations. Azure itself has runtime validation so if we make an API call with an invalid configuration it will fail. The webhooks are a way to shortcircuit that API call so we tell the user right away if they are misconfiguring something. Anything on top of that that can be validated in the webhook is superfluous IMO. The only exception is when we are validating something that requires making an API call itself, for example, if a certain VM size has certain capabilities in a specific Azure region.

Reply:
We do it differently but I completely understand your points.
I've removed the duplicated validation at runtime.

Reply:
@damdo I still see the validation lines in this file, did you push your updates?